### PR TITLE
Convert Text Input to functional widget

### LINKED ISSUE #1039: Convert text input to functional widget
**Enhancement** 

Convert the text input into a functional widget.

### PR DESCRIPTION
**Pull request by nicknisi**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [x] For new widgets, an entry has been added to the `.dojorc`
* [ ] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [x] WidgetProperties are exported

**Description:**

Resolves #1039

### CODE REVIEW

File: src/text-input/index.tsx [hunk 1]
```diff
@@ -23,12 +21,8 @@ export type TextInputType =
 	| 'url'
 	| 'date';
 
-interface TextInputInternalState {
-	dirty?: boolean;
-}
-
 export interface BaseInputProperties<T extends { value: any } = { value: string }>
-	extends ThemedProperties,
+	extends ThemeProperties,
```
Comment (rorticus) [inline src/text-input/index.tsx]: I don't think we need to extend `ThemedProperties` or `FocusProperties` anymore

File: src/text-input/index.tsx [hunk 2]
```diff
@@ -106,46 +100,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
...
 
-	private _onKeyDown(event: KeyboardEvent) {
-		event.stopPropagation();
-		this.properties.onKeyDown &&
-			this.properties.onKeyDown(event.which, () => {
-				event.preventDefault();
-			});
-	}
+	const themeCss = theme.classes(css);
+	const _uuid = icache.getOrSet('uuid', uuid());
```
Comment (rorticus) [inline src/text-input/index.tsx]: Would this be better as `icache.getOrSet('uuid', () => uuid());`? I don't know how expensive it is to create a UUID, but this is going to run every render as-is.

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-450f8an2p.now.sh
* Tests Deployment: https://dojowidgets-inf0nsntn.now.sh

File: src/text-input/index.tsx [hunk 3]
```diff
@@ -106,46 +100,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
...
+		valid: validValue = { valid: undefined, message: '' },
+		widgetId = _uuid
+	} = properties();
+
+	const pattern = patternValue instanceof RegExp ? patternValue.source : patternValue;
 
-	private _callOnValidate(valid: boolean | undefined, message: string) {
-		let { valid: previousValid } = this.properties;
+	function _callOnValidate(valid: boolean | undefined, message: string) {
+		let previousValid: typeof validValue | undefined = validValue;
```
Comment (rorticus) [inline src/text-input/index.tsx]: Should we re-read properties from here? Like, `const { valid } = properties();`. I think it's a no-no to use properties outside the render function.

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-fbih4ykl5.now.sh
* Tests Deployment: https://dojowidgets-inf0nsntn.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-3tj0j1laa.now.sh
* Tests Deployment: https://dojowidgets-inf0nsntn.now.sh

**Comment by tomdye** [comment]
looks like there's an issue with the `extraClasses` usage in cell / grid / header due to it no longer being available on text-input

File: src/text-input/index.tsx [hunk 4]
```diff
@@ -106,46 +97,81 @@ function formatAutocomplete(autocomplete: string | boolean | undefined): string
...
-@diffProperty('trailing', reference)
-export class TextInput extends ThemedMixin(FocusMixin(WidgetBase))<TextInputProperties> {
-	private _onInput(event: Event) {
-		event.stopPropagation();
-		this.properties.onValue &&
-			this.properties.onValue((event.target as HTMLInputElement).value);
-	}
+export const TextInput = factory(function TextInput({
+	middleware: { icache, theme, validity, focus, diffProperty, invalidator },
+	properties
```
Comment (tomdye) [inline src/text-input/index.tsx]: can you add the `id` here and use it to generate the id base as per https://github.com/dojo/widgets/blob/master/src/menu/index.tsx#L289.
You can then remove `uuid`.

Comment (nicknisi) [reply src/text-input/index.tsx]: Good call!

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-cee33360r.now.sh
* Tests Deployment: https://dojowidgets-okxz9k23n.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-fs0qqji4c.now.sh
* Tests Deployment: https://dojowidgets-okxz9k23n.now.sh

File: src/grid/Cell.tsx [hunk 5]
```diff
@@ -106,10 +106,11 @@ export default class Cell extends ThemedMixin(FocusMixin(WidgetBase))<CellProper
 					? w(TextInput, {
 							key: 'input',
 							theme,
-							classes,
+							classes: classes || {
```
Comment (tomdye) [inline src/grid/Cell.tsx]: You're going to need to compose these together in all cases. Ideally we'd use `theme.compose`, but not available for class based widgets.

I believe you just need to merge at the `@dojo/widgets/text-input` level, something like:

```ts
const { classes = {} } = properties();
const passedInputClasses = classes['@dojo/widgets/text-input'] || {};
const newInputClasses = { 
   ...classes, 
   '@dojo/widgets/text-input': {
      ...passedInputClasses,
      input: [ this.theme(css.input) ]
   }
}
```

Comment (nicknisi) [reply src/grid/Cell.tsx]: Good call! I wasn't sure how best to handle the situation and I really didn't want to convert the whole grid in this ticket 😂

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-52ctc9fxf.now.sh
* Tests Deployment: https://dojowidgets-a32mv9kdl.now.sh

**Comment by dojo-site** [comment]
This pull request has been deployed to:
* Docs Deployment: https://dojowidgets-7c3gima7t.now.sh
* Tests Deployment: https://dojowidgets-5kd6b9i70.now.sh

**Pull request merged**